File: LeetCode/ProblemSet/006.ZigZag Conversion/zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
    if (1 === numRows) {
        return s;
    }

    let array = new Array(numRows);

    let num = 0, d = 1; //d = 1 =》 向下，d = -1 =》 斜上
    for (let i = 0, len = s.length; i < len; i++) {
        array[num] = (array[num] ? array[num] : '') + s[i]
        num += d
        if (num === numRows - 1) {
            d = -1
        } else if (num === 0) {
            d = 1
        }
    }
    return array.join('')
};

console.log(convert('LEETCODEISHIRING', 3));
console.log(convert('LEETCODEISHIRING', 4));
console.log(convert('PAYPALISHIRING', 3));
// console.log(convert('PAYPALISHIRING', 4));

File: LeetCode/Go-Solutions/ProblemSet/0006._ZigZag_Conversion/zigzag-conversion.go
package _006__ZigZag_Conversion

import (
	"strings"
)

func convert(s string, numRows int) string {
	if numRows == 1 {
		return s
	}

	slice := make([]string, numRows)

	num, d := 0, 1; //d = 1 =》 向下，d = -1 =》 斜上
	for _, v := range s {
		slice[num] = slice[num] + string(v)
		num += d
		if num == numRows-1 {
			d = -1
		} else if num == 0 {
			d = 1
		}
	}
	return strings.Join(slice, "")
}
